bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -49,13 +50,58 @@ export stream_to_column = (col, new_col, rollover) ->
   return concat_typed_arrays(col, tmp)
 
 # exported for testing
-export patch_to_column = (col, patch) ->
-  for i in [0...patch.length]
-    [ind, value] = patch[i]
-    col[ind] = value
+export slice = (ind, length) ->
+  if isObject(ind)
+    return [ind.start ? 0, ind.stop ? length, ind.step ? 1]
+  return [start, stop, step] = [ind, ind+1, 1]
 
+# exported for testing
+export patch_to_column = (col, patch, shapes) ->
```

Comment:
The slices get turned into `[number, number, number]` inside this function. In the patch that is passed in they are objects: `{start: 0, stop: 10, step: 1}`

But I don't think what you have is quite correct. Here is it put another way: 

* A `patch` is a list of "tuples", where each tuple is `(index, new_data)`
* `new_data` is probably `any` (or `any[]`?)
* However `index` can either be:

- a single number
- a single slice object (see above)
- a length 2 or length three list, where the first element is always a number, and the subsequent elements are either numbers or slice objects


Shapes can also be different lengths. The length is dependent on the corresponding item in the CDS column. I.e. if its "lists of 1d" like for multiline, then individual shapes are just `[number]` but if its "lists of 2d" like images then individual shapes are `[number, number]`